sharding/proposer/service_test.go

```diff
@@ -0,0 +1,261 @@
+	}
+	if *collation.ProposerAddress() != node.Account().Address {
+		t.Errorf("Incorrect proposer address, got %v", *collation.ProposerAddress())
+	}
+	if collation.Header().Sig() != nil {
+		t.Errorf("Proposer signaure can not be empty")
+	}
+}
+
+func TestAddCollation(t *testing.T) {
+	backend, smc, err := setup()
+	if err != nil {
+		t.Fatalf("Failed to deploy SMC contract: %v", err)
+	}
+	node := &mockNode{smc: smc, t: t, backend: backend}
+	var txs []*types.Transaction
+	for i := 0; i < 10; i++ {
+		data := make([]byte, 1024)
+		rand.Read(data)
+		txs = append(txs, types.NewTransaction(0, common.HexToAddress("0x0"),
+			nil, 0, nil, data))
+	}
+
+	collation, err := createCollation(node, big.NewInt(0), big.NewInt(1), txs)
+	if err != nil {
+		t.Errorf("Create collation failed: %v", err)
+	}
+
+	// fast forward to next period.
+	for i := 0; i < int(sharding.PeriodLength); i++ {
+		backend.Commit()
+	}
+
+	// normal test case #1 create collation with normal parameters.
+	err = addHeader(node, *collation)
+	if err != nil {
+		t.Errorf("%v", err)
+	}
+	backend.Commit()
+
+	// verify collation was correctly added from SMC.
+	collationFromSMC, err := smc.CollationRecords(&bind.CallOpts{}, big.NewInt(0), big.NewInt(1))
+	if err != nil {
+		t.Errorf("Failed to get collation record")
+	}
+	if collationFromSMC.Proposer != node.Account().Address {
+		t.Errorf("Incorrect proposer address, got %v", *collation.ProposerAddress())
+	}
+	if common.BytesToHash(collationFromSMC.ChunkRoot[:]) != *collation.Header().ChunkRoot() {
+		t.Errorf("Incorrect chunk root, got %v", collationFromSMC.ChunkRoot)
+	}
+
+	// negative test case #1 create the same collation that just got added to SMC.
+	collation, err = createCollation(node, big.NewInt(0), big.NewInt(1), txs)
+	if err == nil {
+		t.Errorf("Create collation should fail due to same collation in SMC")
+	}
+
+}
+
+func TestCheckCollation(t *testing.T) {
+	backend, smc, err := setup()
+	if err != nil {
+		t.Fatalf("Failed to deploy SMC contract: %v", err)
+	}
+	node := &mockNode{smc: smc, t: t, backend: backend}
+	var txs []*types.Transaction
+	for i := 0; i < 10; i++ {
+		data := make([]byte, 1024)
+		rand.Read(data)
+		txs = append(txs, types.NewTransaction(0, common.HexToAddress("0x0"),
+			nil, 0, nil, data))
+	}
+
+	collation, err := createCollation(node, big.NewInt(0), big.NewInt(1), txs)
+	if err != nil {
+		t.Errorf("Create collation failed: %v", err)
+	}
+
+	for i := 0; i < int(sharding.PeriodLength); i++ {
+		backend.Commit()
+	}
+
+	err = addHeader(node, *collation)
+	if err != nil {
+		t.Errorf("%v", err)
+	}
+	backend.Commit()
+
+	// normal test case 1: check if we can still add header for period 1, should return false.
+	a, err := checkHeaderAvailability(node, big.NewInt(0), big.NewInt(1))
+	if err != nil {
+		t.Errorf("Can not check header availability: %v", err)
+	}
+	if a {
+		t.Errorf("Check header avability should return: %v", !a)
```

Comment:
Also, we know that it should return a falsey value. Wouldn't it be better to report the actual value?

Reply:
Ah. Good catch!